apps/rush-lib/src/logic/taskRunner/ProjectBuilder.ts

```diff
@@ -210,16 +210,25 @@ export class ProjectBuilder extends BaseBuilder {
 
       let projectBuildCache: ProjectBuildCache | undefined;
       if (this._buildCacheConfiguration) {
-        const projectBuildCacheConfiguration: ProjectBuildCacheConfiguration = ProjectBuildCacheConfiguration.loadForProject(
+        const projectBuildCacheConfiguration:
+          | ProjectBuildCacheConfiguration
+          | undefined = await ProjectBuildCacheConfiguration.tryLoadForProjectAsync(
           this._rushProject,
-          this._buildCacheConfiguration
+          terminal
         );
-        projectBuildCache = this._buildCacheConfiguration.cacheProvider.tryGetProjectBuildCache(terminal, {
-          projectBuildCacheConfiguration: projectBuildCacheConfiguration,
-          command: this._commandToRun,
-          projectBuildDeps: projectBuildDeps,
-          packageChangeAnalyzer: this._packageChangeAnalyzer
-        });
+        if (projectBuildCacheConfiguration) {
+          projectBuildCache = this._buildCacheConfiguration.cacheProvider.tryGetProjectBuildCache(terminal, {
+            projectBuildCacheConfiguration: projectBuildCacheConfiguration,
+            command: this._commandToRun,
+            projectBuildDeps: projectBuildDeps,
+            packageChangeAnalyzer: this._packageChangeAnalyzer
+          });
+        } else {
+          terminal.writeVerboseLine(
+            'Project does not have a build-cache.json configuration file, or one provided by a rig, ' +
+              'so it does not support caching.'
+          );
```

Comment:
Here's a couple interesting edge cases:
- What if a project has no build script?  (`"build": ""` in package.json)  Will the cache ignore it?  It should have no impact on downstream projects.
- What if a project DOES have a build script, but the script does not produce any output.  For example, we have an NPM package that contains a bunch of .json files.  The `"build"` script tests them against a JSON schema, and will report an error if they fail, but it does not create any new files.

Reply:
>What if a project has no build script? ("build": "" in package.json) Will the cache ignore it? It should have no impact on downstream projects.

When a project has no build script, we short-circuit out of this function early [here](https://github.com/microsoft/rushstack/blob/master/apps/rush-lib/src/logic/taskRunner/ProjectBuilder.ts#L250). This causes the code that creates cache entries to get skipped. So when this function executes, it'll first query the cache, find nothing (because another run of this function won't create a cache entry), and then return at that short-circuit.

This is sub-optimal, but I've been waiting to do a major refactor of how all of this works as part of the multi-phase work, and I want to do that after we've made a release with this feature (including these follow-up PRs). So - not great, but I want to improve it soon.

>What if a project DOES have a build script, but the script does not produce any output. For example, we have an NPM package that contains a bunch of .json files. The "build" script tests them against a JSON schema, and will report an error if they fail, but it does not create any new files.

If caching is configured, it'll create a cache entry containing no output assets (i.e. - an empty tarball). That seems like the expected behavior.